File: scripts/SendStringMarkers.py
"""Example program to demonstrate how to send string-valued markers into LSL."""

import random
import time

from pylsl import StreamInfo, StreamOutlet

# first create a new stream info (here we set the name to MyMarkerStream,
# the content-type to Markers, 1 channel, irregular sampling rate,
# and string-valued data) The last value would be the locally unique
# identifier for the stream as far as available, e.g.
# program-scriptname-subjectnumber (you could also omit it but interrupted
# connections wouldn't auto-recover). The important part is that the
# content-type is set to 'Markers', because then other programs will know how
#  to interpret the content

#Defining variables to define stream information 
stream_name = 'MyMarkerStream'
stream_type = 'Markers'
number_of_channels = 1
#A value of 0 is made for the sampling rate
# to signify an irregular sampling rate
sampling_rate = 0
data_type = 'string'
unique_id = 'myuidw43536'

#Defining the stream
info = StreamInfo(stream_name, stream_type,
                  number_of_channels,
                  sampling_rate,
                  data_type, unique_id)

# next make an outlet
outlet = StreamOutlet(info)

print("now sending markers...")
# Defining the marker strings
markernames = ['Test', 'Blah', 'Marker',
               'XXX', 'Testtest', 'Test-1-2-3']

#Infinte while loop
while True:
    # pick a sample to send an wait for a bit
    # The function random.choice() picks one of the
    # markers at random from the markernames list.
    # Please note that we push an event marker
    # as a list.
    outlet.push_sample([random.choice(markernames)])
    #We then wait for a random time interval from
    # 1 to 3 seconds. The random.random()function
    # randomly picks a number from 0 to 1. This
    # number is then multiplied by 3. We use time.sleep
    # to wait for that specified time.
    time.sleep(random.random()*3)

File: scripts/SendStringMarkers.m
% instantiate the library
disp('Loading library...');
lib = lsl_loadlib();

% make a new stream outlet
% the name (here MyMarkerStream) is visible to the experimenter and should be chosen so that 
% it is clearly recognizable as your MATLAB software's marker stream
% The content-type should be Markers by convention, and the next three arguments indicate the 
% data format (1 channel, irregular rate, string-formatted).
% The so-called source id is an optional string that allows for uniquely identifying your 
% marker stream across re-starts (or crashes) of your script (i.e., after a crash of your script 
% other programs could continue to record from the stream with only a minor interruption).

%Defining LSL info variables:
%This variable will be shown as the stream name
%in Lab Recorder
stream_name = 'MyMarkerStream';
stream_type = 'Markers';
number_of_channels = 1;
sampling_rate = 0;
data_type = 'cf_string';
unique_id = 'myuniquesourceid23443';

disp('Creating a new marker stream info...');

% Defining LSL meta-information
info = lsl_streaminfo(lib,stream_name,stream_type,...
    number_of_channels, sampling_rate,...
    data_type,unique_id);

% Opening the LSL outlet
disp('Opening an outlet...');
outlet = lsl_outlet(info);

%Defining markers
markers = {'Test', 'Blah', 'Marker', 'XXX', 'Testtest', 'Test-1-2-3'};

% send markers into the outlet
disp('Now transmitting data...');

% This is an infinite loop
while true
    % The loop will pause from a time of 0 to 3 seconds. This is 
    % determined by the function rand() that chooses a number
    % between 0 and 1 which is multiplied by 3. 
    pause(rand()*3);
    
    % This is a way to randomly pick a marker from the markers cell array
    % defined above. It does this by using the function min() to get the 
    % minimum number between two numbers. These two numbers are the 
    % length of the marker cell array (which is 6) and a number between
    % 0 to 6.
    mrk = markers{min(length(markers), 1+floor(rand()*(length(markers))))};
    disp(['now sending ' mrk]);
    % Note that the string is wrapped into a cell-array
    % Pushing a marker to the network.
    outlet.push_sample({mrk});   
end
